tests/h/views/api_auth_test.py

```diff
@@ -15,6 +17,106 @@
+
+        with pytest.raises(httpexceptions.HTTPBadRequest) as exc:
+            auth_ctrl.get()
+
+        assert 'Unsupported response type' in exc.value.message
+
+    def test_get_redirects_if_user_not_logged_in(self, auth_ctrl,
+                                                 pyramid_config, pyramid_request):
+        pyramid_config.testing_securitypolicy(None)
+        pyramid_request.url = 'http://example.com/auth?client_id=bar'
+
+        with pytest.raises(httpexceptions.HTTPFound) as exc:
+            auth_ctrl.get()
+
+        assert exc.value.location == 'http://example.com/login?next={}'.format(
+                                       url_quote(pyramid_request.url, safe=''))
+
+    def test_get_returns_expected_context(self, auth_ctrl, user_service):
+        ctx = auth_ctrl.get()
+
+        assert ctx == {'username': user_service.fetch.return_value.username,
+                       'client_name': 'Hypothesis',
+                       'client_id': 'valid_id',
+                       'response_type': 'code',
+                       'response_mode': 'web_message',
+                       'state': 'a_random_string'}
+
+    def test_post_returns_auth_code(self, post_auth_ctrl, oauth_service):
+        ctx = post_auth_ctrl.post()
+        assert ctx['code'] == oauth_service.create_grant_token.return_value
+
+    def test_post_sets_origin(self, post_auth_ctrl):
+        ctx = post_auth_ctrl.post()
+        assert ctx['origin'] == 'http://example.com'
+
+    def test_post_returns_state(self, post_auth_ctrl, pyramid_request):
+        ctx = post_auth_ctrl.post()
+        assert ctx['state'] == 'a_random_string'
+
+    @pytest.fixture
+    def auth_ctrl(self, factories, pyramid_config, pyramid_request):
+        pyramid_config.testing_securitypolicy('acct:fred@example.org')
+
+        pyramid_request.GET['client_id'] = 'valid_id'
+        pyramid_request.GET['response_mode'] = 'web_message'
+        pyramid_request.GET['response_type'] = 'code'
+        pyramid_request.GET['state'] = 'a_random_string'
+
+        return views.OAuthAuthorizeController(pyramid_request)
+
+    @pytest.fixture
+    def post_auth_ctrl(self, factories, pyramid_config, pyramid_request):
+        pyramid_config.testing_securitypolicy('acct:fred@example.org')
+
+        pyramid_request.POST['client_id'] = 'valid_id'
+        pyramid_request.POST['response_mode'] = 'web_message'
+        pyramid_request.POST['response_type'] = 'code'
+        pyramid_request.POST['state'] = 'a_random_string'
+
+        return views.OAuthAuthorizeController(pyramid_request)
+
+    @pytest.fixture
+    def oauth_service(self, pyramid_config, pyramid_request):
+        svc = mock.Mock(spec_set=oauth_service_factory(None, pyramid_request))
+        pyramid_config.register_service(svc, name='oauth')
+        return svc
+
+    @pytest.fixture
+    def user_service(self, pyramid_config, pyramid_request):
+        svc = mock.Mock(spec_set=user_service_factory(None, pyramid_request))
```

Comment:
Would [`create_autospec`](http://www.voidspace.org.uk/python/mock/helpers.html?highlight=create_autospec#mock.create_autospec) work here? (And also for `oauth_service` above.) Something like:

```python
svc = mock.create_autospec(h.services.user.UserService, instance=True)
```

I think the differences would be that:

1. `create_autospec` recursively mocks the `UserService` class's methods as well so that the tests will crash if the code tries to call a method with the wrong number of arguments, or with a keyword argument that doesn't exist. (`spec_set` prevents calling a method that doesn't exist, but it doesn't prevent passing args that don't exist.)

2. You won't need to call `user_service_factory()` (rather, just pass the `UserService` class to `create_autospec()`), so you won't need to pass in the `pyramid_request` object. Simplifies the fixture code but also the passing `pyramid_request` into `oauth_service_factory()` here is a little misleading because, unless you look very closely, it looks as if the mock `OAuthService` that we're creating actually depends on that particular `pyramid_request` obj that's passed into it (e.g. whether that request is a GET or POST, what params it has), but I don't think it actually does depend on it.

3. You won't be instantiating a real `UserService` object, so these `OAuthAuthorizeController` unit tests won't crash because of bugs in `UserService`'s `__init__()` method.

If I understand your code correctly you're calling `user_service_factory()` in order to instantiate a real `UserService` object so you can pass that real `UserService` object to `spec_set` in order to create a mock user service object to use in the tests. And you need a `pyramid_request` object (with `.authority and `.db`) as  argument because the real `user_service_factory()` code will crash without it.